src/shm_manager.cc

```diff
@@ -43,10 +43,27 @@ SharedMemoryManager::SharedMemoryManager(
 
   try {
     if (create) {
-      shm_obj_ = std::make_unique<bi::shared_memory_object>(
-          bi::open_or_create, shm_region_name.c_str(), bi::read_write);
-      shm_obj_->truncate(shm_size);
+      // Remove (if any) and create the region. Truncate the shm region could
+      // occasionally fail, allow up to max_retry for stability.
+      int max_retry = 3;
+      while (true) {
+        try {
+          bi::shared_memory_object::remove(shm_region_name.c_str());
+          shm_obj_ = std::make_unique<bi::shared_memory_object>(
+              bi::create_only, shm_region_name.c_str(), bi::read_write);
+          shm_obj_->truncate(shm_size);
```

Comment:
> When creating a new shm region, the write (truncate) call could occasionally fail due to the created shm region having "No such file or directory".

Doesn't that mean the `shared_memory_object` is not properly constructed? There is no exception being raised from the line above this?

Reply:
It could be, the exception is always raised from line 54 `shm_obj_->truncate(shm_size);`.

Reply:
I don't see anything immediately wrong with the code, but if `shm_obj_->truncate(shm_size)` is throwing an exception that the file doesn't exist, and the construction of the `shared_memory_object` is not throwing an exception, I would imagine the shm region is beging concurrently deleted/removed unexpectedly.
```
template<class CharT>
inline bool shared_memory_object::priv_open_or_create
   (ipcdetail::create_enum_t type, const CharT *filename, mode_t mode, const permissions &perm)
{
...
      case ipcdetail::DoCreate:
         m_handle = ipcdetail::create_new_file(shmfile.c_str(), mode, perm, true);
      break;
...
   //Check for error
   if(m_handle == ipcdetail::invalid_file()){
      error_info err = system_error_code();
      this->priv_close();
      throw interprocess_exception(err);
   }
```

Maybe you can add some printing/logging around the deletion/removal of a shared memory region to catch some info during the exception cases?

Reply:
Do we know why `truncate` fails?

Reply:
I am going to try @rmccorm4 suggestion earlier to run
```
bi::shared_memory_object::remove(shm_region_name.c_str());
shm_obj_ = std::make_unique<bi::shared_memory_object>(bi::create_only, shm_region_name.c_str(), bi::read_write);
shm_obj_->truncate(shm_size);
```
in a loop with multiple threads, and see if that would break too.

Reply:
Further testing with some printout statements in boost show the shm region was created successfully, which the `::fchmod(m_handle, unix_perm);` returned 0 and the constructor did not throw any error. The error is from the `posix_fallocate()` inside boost `truncate()`:
```
inline void shared_memory_object::truncate(offset_t length)
{
   #ifdef BOOST_INTERPROCESS_POSIX_FALLOCATE
   int ret = posix_fallocate(m_handle, 0, length);

   if (ret && ret != EOPNOTSUPP){
      error_info err(system_error_code());
      throw interprocess_exception(err);
   }
   //ftruncate fallback
   #endif //BOOST_INTERPROCESS_POSIX_FALLOCATE
   if(0 != ftruncate(m_handle, length)){
      error_info err(system_error_code());
      throw interprocess_exception(err);
   }
}
```
Ref: https://www.boost.org/doc/libs/1_76_0/boost/interprocess/shared_memory_object.hpp

Notice the error message "No such file or directory" from boost regarding `posix_fallocate()` is incorrect, because `posix_fallocate()` does not set `errno` https://man7.org/linux/man-pages/man3/posix_fallocate.3.html#RETURN_VALUE, but boost checks the error from `system_error_code()` which actually checks the `errno`:
```
inline int system_error_code() // artifact of POSIX and WINDOWS error reporting
{
   #if defined (BOOST_INTERPROCESS_WINDOWS)
   return winapi::get_last_error();
   #else
   return errno; // GCC 3.1 won't accept ::errno
   #endif
}
```
Ref: https://www.boost.org/doc/libs/1_76_0/boost/interprocess/errors.hpp

The actual error is within the return value which is 4, and according to the code printout from the build environment: 
```
EBADF: 9
EFBIG: 27
EINTR: 4
EINVAL: 22
ENODEV: 19
ENOSPC: 28
EOPNOTSUPP: 95
ESPIPE: 29
```
that 4 matches EINTR, and according to https://man7.org/linux/man-pages/man3/posix_fallocate.3.html#ERRORS, the actual error is "A signal was caught during execution"